File: windows/net/samples.vb/CaptureDS/SampleCallbacks.vb
Imports System
Imports System.IO
Imports System.Runtime.InteropServices
Imports System.Windows.Forms
Imports System.Diagnostics
Imports DirectShowLib
Imports PrimoSoftware.AVBlocks

Namespace CaptureDS
	Friend Class SampleGrabberCB
		Implements ISampleGrabberCB

		Protected lastMediaTime As Long = -1

		Protected name As String
        Protected _mediaState As MediaState
        Protected bProcess As Boolean = True

        ' how many times the callback has been called
        Protected _sampleIndex As Long

        ' how many samples are processed (based on the media time)
        Protected sampleProcessed As Long

        ' how many samples are dropped (based on the media time)
        Protected sampleDropped As Long

        Protected _streamNumber As Integer

        Protected _mainWindow As IntPtr

        Private sampleBuffer() As Byte

        Private Function ProcessSample(ByVal pBuffer As IntPtr, ByVal dataLen As Integer, ByVal sampleTime As Double) As Boolean
            Dim pushResult As Boolean = True

            If (_mediaState IsNot Nothing) AndAlso ((_mediaState.transcoder IsNot Nothing)) Then
                If sampleTime < 0 Then
                    sampleTime = 0
                End If

                If (Nothing Is sampleBuffer) OrElse (sampleBuffer.Length <> dataLen) Then
                    sampleBuffer = New Byte(dataLen - 1) {}
                End If

                Marshal.Copy(pBuffer, sampleBuffer, 0, dataLen)

                Dim inputSample As New MediaSample()
                inputSample.Buffer = New MediaBuffer(sampleBuffer)
                inputSample.StartTime = sampleTime

                ' System.Diagnostics.Debug.WriteLine(string.Format("transcoder.Push(stream: {0}, sampleTime: {1}, sampleData: {2})", StreamNumber, inputSample.StartTime, inputSample.Buffer.DataSize));

                ' transcoder.Push() is not threads safe.
                ' lock (transcoder){ } ensure that only one thread is calling transcoder.Push()
                SyncLock _mediaState.transcoder
                    pushResult = _mediaState.transcoder.Push(StreamNumber, inputSample)
                End SyncLock
            End If

            If pushResult Then
                Return True
            End If

            WinAPI.PostMessage(MainWindow, Util.WM_STOP_CAPTURE, New IntPtr(_streamNumber), IntPtr.Zero)
            bProcess = False
            Return False
        End Function


        Public Function BufferCB(ByVal sampleTime As Double, ByVal pBuffer As IntPtr, ByVal bufferLen As Integer) As Integer Implements ISampleGrabberCB.BufferCB
            If Not bProcess Then
                Return WinAPI.E_FAIL
            End If

            _sampleIndex += 1

            Dim processed As Boolean = ProcessSample(pBuffer, bufferLen, sampleTime)

            Return If(processed, WinAPI.S_OK, WinAPI.E_FAIL)

        End Function

        Public Function SampleCB(ByVal sampleTime As Double, ByVal pSample As IMediaSample) As Integer Implements ISampleGrabberCB.SampleCB
            Try
                If Not bProcess Then
                    Return WinAPI.E_FAIL
                End If

                ' internal stats
                _sampleIndex += 1
                Dim tStart As Long = Nothing, tEnd As Long = Nothing

                pSample.GetMediaTime(tStart, tEnd)
                Debug.Assert(tStart < tEnd)
                Debug.Assert(tStart > lastMediaTime)
                sampleProcessed += tEnd - tStart
                sampleDropped += tStart - lastMediaTime - 1
                lastMediaTime = tEnd - 1

                Dim dataLen As Integer = pSample.GetActualDataLength()
                Dim bufPtr As IntPtr = Nothing
                Dim hr As Integer = pSample.GetPointer(bufPtr)
                Debug.Assert(0 = hr)

                Dim processed As Boolean = ProcessSample(bufPtr, dataLen, sampleTime)

                Return If(processed, WinAPI.S_OK, WinAPI.E_FAIL)
            Catch ex As Exception
                System.Diagnostics.Debug.WriteLine(ex.ToString())
            Finally
                Marshal.ReleaseComObject(pSample)
            End Try

            Return WinAPI.E_FAIL
        End Function ' end of SampleCB

        Public Property MainWindow() As IntPtr
            Get
                Return _mainWindow
            End Get
            Set(ByVal value As IntPtr)
                _mainWindow = value
            End Set
        End Property

        Public ReadOnly Property SampleIndex() As Long
            Get
                Return _sampleIndex
            End Get
        End Property

        Public Property StreamNumber() As Integer
            Get
                Return _streamNumber
            End Get
            Set(ByVal value As Integer)
                _streamNumber = value
            End Set
        End Property

        Public ReadOnly Property ProcessedSamples() As Long
            Get
                Return sampleProcessed
            End Get
        End Property

        Public ReadOnly Property DroppedSamples() As Long
            Get
                Return sampleDropped
            End Get
        End Property

        Public Sub New(ByVal name As String)
            If Not String.IsNullOrEmpty(name) Then
                Me.name = name
            Else
                Me.name = "SampleGrabberCB"
            End If
        End Sub

        Protected Overrides Sub Finalize()
            Reset()
        End Sub

        Public Property MediaState() As MediaState
            Get
                Return _mediaState
            End Get
            Set(ByVal value As MediaState)
                _mediaState = value
            End Set
        End Property

        Public Sub Reset()
            _sampleIndex = 0
            sampleProcessed = 0
            sampleDropped = 0

            _mediaState = Nothing
            bProcess = True

            lastMediaTime = -1

            sampleBuffer = Nothing
        End Sub

	End Class ' end of SampleGrabberCB

	' Sample grabber callback method
	Friend Enum CBMethod
		Sample = 0 ' the original sample from the upstream filter
		Buffer = 1 ' a copy of the sample of the upstream filter
	End Enum
End Namespace

File: windows/net/samples/capture_ds_video_audio/SampleCallbacks.cs
using System;
using System.IO;
using System.Runtime.InteropServices;
using System.Windows.Forms;
using System.Diagnostics;
using DirectShowLib;
using PrimoSoftware.AVBlocks;

namespace CaptureDS
{
    class SampleGrabberCB: ISampleGrabberCB
    {
        protected long lastMediaTime = -1;

        protected string name;
        protected MediaState mediaState;
        protected bool bProcess = true;
        
        // how many times the callback has been called
        protected long sampleIndex;

        // how many samples are processed (based on the media time)
        protected long sampleProcessed;

        // how many samples are dropped (based on the media time)
        protected long sampleDropped;

        protected int streamNumber;
        protected IntPtr mainWindow;


        /*
         * Specifies whether to use Transcoder.PushUnmanaged or Transcoder.Push for encoding.
         * PushUnmanaged uses UnmanagedMediaBuffer to pass the captured unamanaged data directly
         * to the Transcoder and saves copying from unmanaged to managed memory.
         */ 
        private bool unmanaged = false;

        MediaSample sample = new MediaSample();

        bool ProcessSample(IntPtr pBuffer, int dataLen, double sampleTime)
        {
            bool pushResult = true;

            if ((mediaState != null) && ((mediaState.transcoder != null)))
            {
                if (sampleTime < 0)
                    sampleTime = 0;

                if (unmanaged)
                {
                    if (sample.UnmanagedBuffer == null)
                        sample.UnmanagedBuffer = new UnmanagedMediaBuffer();

                    sample.UnmanagedBuffer.Attach(pBuffer, dataLen, true);
                }
                else
                {
                    if (sample.Buffer == null || sample.Buffer.Start.Length != dataLen)
                    {
                        sample.Buffer = new MediaBuffer(new byte[dataLen]);
                    }

                    Marshal.Copy(pBuffer, sample.Buffer.Start, 0, dataLen);
                    sample.Buffer.SetData(0, dataLen);
                }

                sample.StartTime = sampleTime;

                //System.Diagnostics.Trace.WriteLine(
                //    string.Format("transcoder.Push(stream: {0}, sampleTime: {1}, sampleData: {2})", 
                //    StreamNumber, sample.StartTime, sample.Buffer.DataSize));

                /*
                 * The built-in transcoder is not thread-safe
                 * and usually the calls to Transcoder.Push/PushUnmanaged must be synchronized,
                 * However the CompositeTranscoder (implemented in this app)
                 * allows that input samples from different streams can be pushed simultaneously.
                 *  
                 */
#if (TRACE && DEBUG)
                lock (mediaState.transcoder)
#endif
                {
                    if (unmanaged)
                    {
                        pushResult = mediaState.transcoder.PushUnmanaged(StreamNumber, sample);
                    }
                    else
                    {
                        pushResult = mediaState.transcoder.Push(StreamNumber, sample);
                    }
                }
            }

            if (pushResult)
                return true;


            WinAPI.PostMessage(MainWindow, Util.WM_STOP_CAPTURE, new IntPtr(streamNumber), IntPtr.Zero);
            bProcess = false;
            return false;
        }


        public int BufferCB(double sampleTime, IntPtr pBuffer, int bufferLen)
        {
            if (!bProcess)
                return WinAPI.E_FAIL;

            sampleIndex += 1;

            bool processed = ProcessSample(pBuffer, bufferLen, sampleTime);

            return processed ? WinAPI.S_OK : WinAPI.E_FAIL;

        }

        public int SampleCB(double sampleTime, IMediaSample pSample)
        {
            try
            {
                if (!bProcess)
                    return WinAPI.E_FAIL;

                // internal stats
                ++sampleIndex;
                long tStart, tEnd;

                pSample.GetMediaTime(out tStart, out tEnd);
                Debug.Assert(tStart < tEnd);
                Debug.Assert(tStart > lastMediaTime);
                sampleProcessed += tEnd - tStart;
                sampleDropped += tStart - lastMediaTime - 1;
                lastMediaTime = tEnd - 1;

                int dataLen = pSample.GetActualDataLength();
                IntPtr bufPtr;
                int hr = pSample.GetPointer(out bufPtr);
                Debug.Assert(0 == hr);

                bool processed = ProcessSample(bufPtr, dataLen, sampleTime);

                return processed ? WinAPI.S_OK : WinAPI.E_FAIL;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine(ex.ToString());
            }
            finally
            {
                Marshal.ReleaseComObject(pSample);
            }

            return WinAPI.E_FAIL;
        } // end of SampleCB

        public IntPtr MainWindow
        {
            get { return mainWindow; }
            set { mainWindow = value; }
        }

        public long SampleIndex
        {
            get { return sampleIndex; }
        }

        public int StreamNumber
        {
            get { return streamNumber; }
            set { streamNumber = value; }
        }

        public long ProcessedSamples
        {
            get { return sampleProcessed; }
        }

        public long DroppedSamples
        {
            get { return sampleDropped; }
        }

        public SampleGrabberCB(string name)
        {
            if (!string.IsNullOrEmpty(name))
            {
                this.name = name;
            }
            else
            {
                this.name = "SampleGrabberCB";
            }
        }

        ~SampleGrabberCB()
        {
            Reset();
        }

        public MediaState MediaState
        {
            get { return mediaState; }
            set { mediaState = value; }
        }

        public void Reset()
        {
            sampleIndex = 0;
            sampleProcessed = 0;
            sampleDropped = 0;

            mediaState = null;
            bProcess = true;

            lastMediaTime = -1;

            sample.Buffer = null;

            if (sample.UnmanagedBuffer != null)
                sample.UnmanagedBuffer.Release();

            sample.UnmanagedBuffer = null;
        }

    } // end of SampleGrabberCB

    // Sample grabber callback method
    enum CBMethod
    {
        Sample = 0, // the original sample from the upstream filter
        Buffer = 1  // a copy of the sample of the upstream filter
    };
}
